src/lib/storage/index/partial_hash/flat_map_iterator.hpp

```diff
@@ -0,0 +1,75 @@
+#pragma once
+
+#include "flat_map_iterator_impl.hpp"
+#include "types.hpp"
+
+/**
+ * The choice of the concrete iterator implementation was subject to a long discussion (see PR 2448). The
+ * PartialHashIndex uses the pointer to implementation (pImpl) idiom, i.e., the non-templated PartialHashIndex instance
+ * holds an opaque pointer to its templated implementation PartialHashIndexImpl. Since the index access is done through
+ * its iterators (more precisely, through its access methods that also work with iterators), a non-templated iterator is
+ * also needed.
+ *
+ * We use the pImpl idiom to implement the FlatMapIterator as a non-templated class. Since FlatMapIteratorImpl is a
+ * class template, we use BaseFlatMapIteratorImpl as a non-templated base class for the FlatMapIteratorImpl so that
+ * FlatMapIterator can hold a pointer to a non-templated BaseFlatMapIteratorImpl instance. The PartialHashIndex's
+ * iterator functions return (pairs of) FlatMapIterators so that PartialHashIndex itself can be non-templated.
+ *
+ *   +-----------------+            +-------------------------+
+ *   |                 |            | <<Abstract>>            |
+ *   | FlatMapIterator |----------->| BaseFlatMapIteratorImpl |
+ *   |                 | unique_ptr |                         |
+ *   +-----------------+            +-------------------------+
+ *                                              ^
+ *                                             /_\
+ *                                              |    .......................
+ *                                              |    : Template parameter: :
+ *                                    +--------------: DataType            :
+ *                                    |              :.....................:
+ *                                    | FlatMapIteratorImpl  |
+ *                                    |                      |
+ *                                    +----------------------+
+ *
+ * The following design attempts were discarded:
+ *
+ * Iterator as a template parameter: Using templated iterators as a template decreased the code quality significantly.
+ * Also, static and dynamic polymorphism do not go well together because virtual functions cannot be templated.
+ *
+ * Returning BaseFlatMapIteratorImpl pointers rather than FlatMapIterator instances: Since the methods of the
+ * FlatMapIteratorImpl must be accessed, it would be necessary to cast the pointers every time the index is accessed.
+ * If the BaseFlatMapIteratorImpl pointer is not casted to FlatMapIteratorImpl, only the methods of the Base instance
```

Comment:
But this is exactly what runtime polymorphism should allow us to do, i.e., calling the derived class's behavior without the need to cast it.